### PR TITLE
Enhancement/nextgen

### PR DESCRIPTION
**Pull request by andreasrs**
- [x] Agree on design direction / look
- [x] Agree on options split (fetch/redux)
- [x] Decide on further iterations in this PR, vs merge + contributions in newer PRs

Features:
- Provide optional data-fetch example (select during init)
- Provide optional redux example (select during init)
- Streamlined design and comply more with next.getroc.org
- First step toward more user-friendly template
- Base for further improvements

**Some screens:**
2016-12-04-183118_672x300_scrot

The template will add/remove menu items as selected during project init.

2016-12-04-184042_828x657_scrot

2016-12-04-184101_812x798_scrot

2016-12-04-184115_826x651_scrot

### CODE REVIEW

**Comment by andreasrs** [comment]
For testing this PR branch locally in a directory called `roctest`, one could use `wget https://github.com/rocjs/roc-template-web-app-react/archive/enhancement/nextgen.zip && roc new roctest ./nextgen.zip`

I could not get the branch-specifier to work, and not had the time to look into it yet.

File: template/src/components/fetching/index.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
+// React Hot Reload does not support stateless function components as of now
+/* eslint-disable react/prefer-stateless-function */
+import React, { Component } from 'react';
+import { provideHooks } from 'redial';
+import isoFetch from 'isomorphic-fetch';
```
Comment (dlmr) [inline template/src/components/fetching/index.js]: I think we should call this `fetch` since that is the standard.

File: template/roc.config.js [hunk 2]
```diff
@@ -5,10 +5,23 @@ module.exports = {
       port: {{ port }},
       serve: ['public', 'build/client'],
       favicon: 'favicon.png',
+      {{#if_eq fetchExample true}}
```
Comment (dlmr) [inline template/roc.config.js]: Using an `#if` here should also work, right?

File: template/package.json [hunk 3]
```diff
@@ -4,19 +4,30 @@
...
-    "roc-package-web-app-react": "^1.0.0-beta.3"
+    "babel-eslint": "^7.1.0",
+    "eslint": "^3.6.0",
+    "eslint-config-airbnb": "^12.0.0",
+    "eslint-plugin-import": "^1.16.0",
+    "eslint-plugin-jsx-a11y": "^2.2.2",
+    "eslint-plugin-react": "^6.3.0",
+    {{#if fetchExample}}
+    "isomorphic-fetch": "^2.2.1",
+    {{else}}
```
Comment (dlmr) [inline template/package.json]: I think an `else if` would look a little better here.

File: template/package.json [hunk 4]
```diff
@@ -4,19 +4,30 @@
   "description": "{{ description }}",
   "author": "{{ author }}",
   "license": "{{ license }}",
-  {{#if_eq license "UNLICENSED"}}
```
Comment (dlmr) [inline template/package.json]: Why is this removed?

File: template/src/components/redux/index.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      });
+    };
+
+    // dispatch the thunk that will later dispatch a success of failure action depending on outcome
+    dispatch(fetchThunk);
+  },
+})
+@connect(state => ({
+  loading: state.repo.loading,
+  error: state.repo.error,
```
Comment (dlmr) [inline template/src/components/redux/index.js]: `error` and `loading` seems to not be used here.

File: template/src/components/redux/index.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    const fetchThunk = (thunkDispatch) => {
+      fetchRocStargazers().then((json) => {
+        thunkDispatch(repoFetchSuccess(repoApiUrl, json));
+      }).catch((err) => {
+        thunkDispatch(repoFetchFail(repoApiUrl, err));
+      });
+    };
+
+    // dispatch the thunk that will later dispatch a success of failure action depending on outcome
+    dispatch(fetchThunk);
```
Comment (dlmr) [inline template/src/components/redux/index.js]: This should be returned so the server waits on the result from the API.

File: template/src/components/fetching/index.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  };
+
+  render() {
+    if (this.props.loading) {
+      return <ul><li>Loading...</li></ul>;
+    }
+
+    let repoItems = [];
+    if (this.props.repo.length > 0) {
+      repoItems = this.props.repo.map(repo => <li key={repo.id}>{repo.name} </li>);
```
Comment (dlmr) [inline template/src/components/fetching/index.js]: What do you think about moving the `display` to the number of elements created here. This would mean that we do not create more items than we need.

File: template/src/components/fetching/index.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+  static defaultProps = {
+    display: 5,
+    repoItems: [],
+    loading: true,
+  };
+
+  render() {
+    if (this.props.loading) {
+      return <ul><li>Loading...</li></ul>;
```
Comment (dlmr) [inline template/src/components/fetching/index.js]: Would it look better here not to use a list for the loading state?

File: template/src/components/fetching/index.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    let repoItems = [];
```
Comment (dlmr) [inline template/src/components/fetching/index.js]: What do you think about naming this something other than what the prop is called, maybe `repoItemsToDisplay` or similar.

**Comment by andreasrs** [comment]
Thanks for the comments @dlmr - do you have any feedback on the conceptual level? Do you think the redux and fetch is a good initial optional split?

I'll go over an clean up the code once that has been decided. Also, did you have an alternative colour-set suggestion?

I also think we should consider a (separate) super-minimalistic template with as few opinions as possible. For those that just want to get started with an empty component.

**Comment by dlmr** [comment]
It is possible to install this directly without the need of using `wget` but it is a little tricky because of a bug in the template logic. I have created a PR that addresses this https://github.com/rocjs/roc/pull/153.

To install directly use the clone path along with the branch name, like this:
```sh
$ roc init git@github.com:rocjs/roc-template-web-app-react.git enhancement/nextgen
```
---
Regarding the conceptual level. I think this is a great start but I would like to see that the "default" version, when answering no to everything or a higher level question, results in an almost empty project with only a single `App.js` or similar. This is my take on what you are describing as _"super-minimalistic template "_. 

One thing that might be confusing for beginners is that there is data fetching in both the __data-fetching example__ and the __Redux example__. One might think that you will not get any data fetching if only selecting Redux. One way of addressing this is to tell the user that data fetching will be included if selecting the Redux example.

File: template/package.json [hunk 10]
```diff
@@ -4,19 +4,30 @@
...
-  {{/if_eq}}
   "scripts": {
     "build": "roc build",
     "dev": "roc dev",
     "start": "roc start"
   },
   "dependencies": {
-    "roc-package-web-app-react": "^1.0.0-beta.3"
+    "babel-eslint": "^7.1.0",
+    "eslint": "^3.6.0",
```
Comment (fix-fix) [inline template/package.json]: It looks like eslint was added as a dependency, but there is no way to run linting other than using `node_modules/.bin/eslint` (directly or in npm scripts) or other tools, that can work with local eslint package (IDEs and text editors linter plugins).

Maybe it's better to move common eslint setup and React-specific configs into its own plugins, providing roc commands for linting source files?

Comment (andreasrs) [reply template/package.json]: 👋 
@dlmr has a local working thing here, which I will introduce once that has been pushed and published ✅ For now I added a npmscript on the branch.

**Comment by andreasrs** [comment]
I think we can squash+merge this now @dlmr 
Then we can update our tutorial on current getroc.org to not use the `wip` branch.

If we want to make the "simple" case even cleaner and more simple, I propose we do it separately in another PR.

File: template/package.json (hunk 4 above)
Comment (andreasrs) [reply template/package.json]: This was a mistake, added it back

**Comment by andreasrs** [comment]
@dlmr ping 🎉

**Review by dlmr — APPROVED** [review]

**Comment by dlmr** [comment]
🎉
